### PR TITLE
Support for geometry digitizing when adding a child feature via relation editor

### PR DESCRIPTION
**Pull request by nirvn**
When adding a feature via the relation editor, if the referenced relation layer has a geometry, the user will jump into digitizing mode to add his/her point/line/polygon.

How it looks:
Peek 2020-12-30 18-53

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7158ea78999ecc2c5334043908b2ccdc98fa905b:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-9aa6d2e191758bdac73854e6166d6f442b9379cf-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-9aa6d2e191758bdac73854e6166d6f442b9379cf-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-9aa6d2e191758bdac73854e6166d6f442b9379cf-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-9aa6d2e191758bdac73854e6166d6f442b9379cf-x86.apk)

File: src/qml/editorwidgets/CheckBox.qml [hunk 1]
```diff
@@ -3,6 +3,7 @@ import QtQuick.Controls 2.12
 
 Item {
   signal valueChanged( var value, bool isNull )
+  signal requestGeometry(var item, var layer)
```
Comment (m-kuhn) [inline src/qml/editorwidgets/CheckBox.qml:6]: Could we let these tools inherit from `EditorWidgetBase` instad of `Item` and put this boilerplate code in there?

File: src/qml/DigitizingToolbar.qml [hunk 2]
```diff
@@ -19,6 +19,7 @@ VisibilityFadingRow {
   spacing: 4
 
   signal confirm
+  signal cancel
```
Comment (m-kuhn) [inline src/qml/DigitizingToolbar.qml:22]: Should we follow the qdoc syntax (https://doc.qt.io/qt-5/qtwritingstyle-qml.html#signals-and-handlers-documentation) for these?
I.e.

>     This signal is emitted when [...] . The corresponding handler is \c onCancel.

File: src/qml/editorwidgets/RelationEditor.qml [hunk 3]
```diff
@@ -101,6 +105,12 @@ Rectangle{
               if( save() ) {
                 //this has to be checked after buffering because the primary could be a value that has been created on creating featurer (e.g. fid)
                 if( relationEditorModel.parentPrimariesAvailable ) {
+                    console.log(relationEditorModel.relation.referencingLayer.geometryType())
+                    if ( relationEditorModel.relation.referencingLayer.geometryType() !== QgsWkbTypes.NullGeometry )
+                    {
+                        requestGeometry(relationEditor, relationEditorModel.relation.referencingLayer);
+                        return;
+                    }
                     embeddedPopup.state = 'Add'
```
Comment (m-kuhn) [inline src/qml/editorwidgets/RelationEditor.qml:114]: Can we move the duplicated code here to a function?

**Comment by m-kuhn** [comment]
Wow, nice work !

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 85dd04318f8045d3e071388e9bc64df5dc31e3be:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-2c6af1ac056234fa014c1a50927cba98be632195-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-2c6af1ac056234fa014c1a50927cba98be632195-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-2c6af1ac056234fa014c1a50927cba98be632195-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-2c6af1ac056234fa014c1a50927cba98be632195-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 1a8f12ea849574cd2fc8198a12ce556744f6e528:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-08d5d032ca3d0948f1d22f805ee4bca8cc588c3a-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-08d5d032ca3d0948f1d22f805ee4bca8cc588c3a-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-08d5d032ca3d0948f1d22f805ee4bca8cc588c3a-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-08d5d032ca3d0948f1d22f805ee4bca8cc588c3a-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 97386f35006357ad413a74b6d8c6150fecf2e334:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-5bd91a6de5f95c906c794c8a227fec1677297e62-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-5bd91a6de5f95c906c794c8a227fec1677297e62-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-5bd91a6de5f95c906c794c8a227fec1677297e62-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-5bd91a6de5f95c906c794c8a227fec1677297e62-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 210dfe695ea1fbfb7f1bb81dcfa53bb882abcd85:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-741b7d978923401cd29466830992ef9ad3c826ca-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-741b7d978923401cd29466830992ef9ad3c826ca-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-741b7d978923401cd29466830992ef9ad3c826ca-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-741b7d978923401cd29466830992ef9ad3c826ca-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for dd4f771c7e91cf8de3eebdbf9fcc582049ceeef7:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b3aaa2cab8138a6a48f1506cdec9b08e119c238c-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b3aaa2cab8138a6a48f1506cdec9b08e119c238c-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b3aaa2cab8138a6a48f1506cdec9b08e119c238c-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b3aaa2cab8138a6a48f1506cdec9b08e119c238c-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e15d8947647b5b3f41c3e2d38bc79295b99c9b9a:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-4f21db246bb85855e8f763e900431eaecfb993b6-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-4f21db246bb85855e8f763e900431eaecfb993b6-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-4f21db246bb85855e8f763e900431eaecfb993b6-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-4f21db246bb85855e8f763e900431eaecfb993b6-x86.apk)

**Comment by jhonalex06** [comment]
I tested the PR and obtained the next results: 

1. Try with a project generated from QGIS Model Baker, this project has the next structure:



In this project, I tried to create a new **Direction** (Point) from **Predio** (Alphanumeric), but always when I did it, the application crashed like this: 

project_1

2. In view of the first case behavior, I try to generate a new project from QGIS similar to the example of the PR 1544 (https://github.com/opengisch/QField/pull/1544).
3.

**Comment by m-kuhn** [comment]
Thanks @jhonalex06 

- Can you provide us with a test project to reproduce the behavior?
- Ideally, could you connect your test device to a computer using a cable and use `adb logcat` while reproducing the crash, that will help a lot to debug. (upload it to gist.github.com and paste the trace here).

**Comment by jhonalex06** [comment]
> Thanks @jhonalex06
> 
>     * Can you provide us with a test project to reproduce the behavior?
> 
>     * Ideally, could you connect your test device to a computer using a cable and use `adb logcat` while reproducing the crash, that will help a lot to debug. (upload it to gist.github.com and paste the trace here).

Thanks @m-kuhn for your reply,

The test projects are here: 
[Supplies.zip](https://github.com/opengisch/QField/files/5851008/Supplies.zip)

And the logcat is here: 
https://gist.github.com/jhonalex06/083b7edd3434619e659441ef73d639fa

**Comment by nirvn** [comment]
@jhonalex06 , thanks for testing this new feature. I've been able to replicate and fix your crash as well as the calendar issue. The latest APKs (which will appear in 10 minutes or so) should make it work smoothly for you.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for cc50833dd19319b264a67f321a754787913ca383:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b613fdaed3bbea42be15a77e4a9f45edc0ab5528-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b613fdaed3bbea42be15a77e4a9f45edc0ab5528-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b613fdaed3bbea42be15a77e4a9f45edc0ab5528-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-b613fdaed3bbea42be15a77e4a9f45edc0ab5528-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for f7ffa88fba34a2f2869b853dfc2f9623cf5b6485:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-a7cde901ad80be2cee3a7d4ff5be0e03ecb9d98b-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-a7cde901ad80be2cee3a7d4ff5be0e03ecb9d98b-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-a7cde901ad80be2cee3a7d4ff5be0e03ecb9d98b-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-a7cde901ad80be2cee3a7d4ff5be0e03ecb9d98b-x86.apk)

**Review by m-kuhn — APPROVED** [review]
Excellent work code wise
Waiting for the final test results from @jhonalex06 before merging

**Comment by jhonalex06** [comment]
> Excellent work code wise
> Waiting for the final test results from @jhonalex06 before merging

I tested the PR with the same structure as the project_1, but with some modifications in the forms, this project has the name project_3, and obtained the next results:

This PR (1544) generates changes in the forms, for example, when using the widget Value Relation, on QField 1.81 the widget works as expected, but on this PR, the widget Value Relation disappears.

Expect behavior (QField 1.81)

image

Observed behavior (QField PR 1544)

image

For this reason, when I tried to create a new direction (point geometry), the form appears without widgets like this:

I also found a bug in the widget Value Map, in this case when I accessed a child feature the widget Value Map doesn’t work, I clicked but it doesn’t show the attributes. 

Note: All the examples were tested on the cell phone Moto G5 Plus.

A test project to reproduce the behavior is this:
[project_3.zip](https://github.com/opengisch/QField/files/5867866/project_3.zip)

@nirvn, @m-kuhn, Thanks for the fixes and I will be waiting to test the new changes.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 8af559d6b924762d13995bb9ca10392632566348:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-31078d67657207516bf32aa84267d774c74d8aad-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-31078d67657207516bf32aa84267d774c74d8aad-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-31078d67657207516bf32aa84267d774c74d8aad-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-31078d67657207516bf32aa84267d774c74d8aad-x86.apk)

**Comment by nirvn** [comment]
@jhonalex06 , the broken state of the value relationship wasn't linked to this PR's changes, but well spotted. It's been fixed. The other issue, missing combobox popup list in embedded child feature form has also been fixed. Give the APKs a try.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 920452733b9c88aa15b7d062c38c8a5c9c09b21d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-17201d70bc7b5599369e58dc05cb32d3000f6187-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-17201d70bc7b5599369e58dc05cb32d3000f6187-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-17201d70bc7b5599369e58dc05cb32d3000f6187-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-17201d70bc7b5599369e58dc05cb32d3000f6187-x86.apk)

**Comment by jhonalex06** [comment]
> @jhonalex06 , the broken state of the value relationship wasn't linked to this PR's changes, but well spotted. It's been fixed. The other issue, missing combobox popup list in embedded child feature form has also been fixed. Give the APKs a try.

The project_3 has a relation between **Predio** (Alphanumeric) -> **construccion** (Alphanumeric) -> **Unidad de construccion** (Polygon). When I tested this relation, I found some bugs. First, the form of **Unidad de construccion** (Polygon) appears after the form construccion (Alphanumeric) and this can be confusing for users because in other cases it works different. For instance, in  **Predio** (Alphanumeric) -> **Dirección** (Point), when I create a new **Dirección** (Point), the first form is **Dirección** (Point), then when I finish the **Dirección** form, the **Predio** form comes out.

project-3-ene-26

I also found a bug when I finish filling the fields of the relation **Predio** (Alphanumeric) -> **construccion** (Alphanumeric)-> **Unidad de construccion** (Polygon), I found that the checkmark icon continues to show in the interface without a reason.
 
image

Note: All the examples were tested on the cell phone Moto G5 Plus.

A test project to reproduce the behavior is this: 
[project_3.zip](https://github.com/opengisch/QField/files/5874173/project_3.zip)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7c2850f9df2558a90b9701ae272c84bbbd334599:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0c270475c56cae100e835366aa031cc63178f874-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0c270475c56cae100e835366aa031cc63178f874-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0c270475c56cae100e835366aa031cc63178f874-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0c270475c56cae100e835366aa031cc63178f874-x86.apk)

**Comment by nirvn** [comment]
@jhonalex06 , latest pair of issues fixed. Your complex form setup is a great stress test.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for cf2b0b98e90dec25e09ed944096198256b54d1d2:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-c14b1a84e82ae77534fbcbd051378aac4e930bfd-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-c14b1a84e82ae77534fbcbd051378aac4e930bfd-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-c14b1a84e82ae77534fbcbd051378aac4e930bfd-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-c14b1a84e82ae77534fbcbd051378aac4e930bfd-x86.apk)

**Comment by jhonalex06** [comment]
> @jhonalex06 , latest pair of issues fixed. Your complex form setup is a great stress test.

Yes @nirvn , I agree, this form setup is a great stress test 😉.

I tested and this PR works pretty well 👍🏿 , Thanks for the fixes

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 92f816e927d20112010362faa3ec1a8f4b2569dc:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0d3131f6fca4917ea2a6b3795353eec5c09be026-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0d3131f6fca4917ea2a6b3795353eec5c09be026-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0d3131f6fca4917ea2a6b3795353eec5c09be026-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1544-0d3131f6fca4917ea2a6b3795353eec5c09be026-x86.apk)

**Pull request merged**